[Search Map] Fix searchThisArea

DieBatzen opened the pull request:
#2708
Fix function 'searchThisArea' after GS tech migration

2Abendsegler commented:
@DieBatzen 
Sorry wegen der langen Wartezeit. Ich war ne Weile ausgefallen. Ich versuchs in den nächsten Tagen zu prüfen.

2Abendsegler commented:
Moin,
es scheint noch ein paar Probleme zu geben - scheiß Thema. 🙄

1. Unter Firefox, scheinbar nicht unter Chrome, gibt es bei mir ziemlich häufig Fehlermeldungen, aber nicht immer. Diese Fehler entstehen nicht nur durch deine letzten Änderungen, sondern auch schon beim Setzen der Default Filter. Letzteres soll aber im Moment noch kein Thema sein.
Hier mal eine Fehlermeldung exemplarisch. Das ist nur eine Fehlermeldung, mein Bildschirm am Laptop ist aber nicht groß genug für einen Screenshot:
![1](https://github.com/user-attachments/assets/cb8439eb-a27c-48e5-aad7-09c3c1f40e6a)
![2](https://github.com/user-attachments/assets/65c5289f-8afd-4275-a2aa-990d9f171488)
Und hier die entsprechende Programmstelle:
![3](https://github.com/user-attachments/assets/17db0dad-b12e-4f98-b5c2-96ba90092c80)
`// Check 1min limit.`

2. Zoomt man hinein mit dem Mausrad werden die Cache Icons nicht aktualisiert. Die Koordinaten werden dabei im Suchfeld in der Sidebar auch nicht aktualisiert. Zoomt man heraus funktioniert beides. Das gilt für Firefox und für Chrome. Dieser Fehler zeigt sich bei mir nicht immer, aber in den meisten Fällen. Ist der Fehler einmal aufgetreten, scheint er in der Folge immer weiter aufzutreten. Ich kann im Moment nicht beurteilen, ob das in der alten Version auch schon fehlerhaft war. 
Ich habe dazu mal ein gif aufgenommen, in dem ich je zweimal hinein und heraus zoome. Ich hatte auch schon viel länger gewartet nach dem Zoomen um sicherzustellen, dass es wirklich nicht funktioniert. Für die Aufnahme habe ich es aber etwas kürzer gestaltet.
![Zoomen](https://github.com/user-attachments/assets/69f83c8d-951b-4852-a150-62aa2fbc7daa)

DieBatzen commented:
> Unter Firefox, scheinbar nicht unter Chrome, gibt es bei mir ziemlich häufig Fehlermeldungen, aber nicht immer. Diese Fehler entstehen nicht nur durch deine letzten Änderungen, sondern auch schon beim Setzen der Default Filter. Letzteres soll aber im Moment noch kein Thema sein.

Das hängt wahrscheinlich hiermit zusammen: [https://github.com/2Abendsegler/GClh/issues/2708#issuecomment-2704763097](https://github.com/2Abendsegler/GClh/issues/2708#issuecomment-2704763097)
Schaue ich mir an.

DieBatzen commented:
> Zoomt man hinein mit dem Mausrad werden die Cache Icons nicht aktualisiert. Die Koordinaten werden dabei im Suchfeld in der Sidebar auch nicht aktualisiert. Zoomt man heraus funktioniert beides. Das gilt für Firefox und für Chrome. Dieser Fehler zeigt sich bei mir nicht immer, aber in den meisten Fällen. Ist der Fehler einmal aufgetreten, scheint er in der Folge immer weiter aufzutreten. Ich kann im Moment nicht beurteilen, ob das in der alten Version auch schon fehlerhaft war.
Ich habe dazu mal ein gif aufgenommen, in dem ich je zweimal hinein und heraus zoome. Ich hatte auch schon viel länger gewartet nach dem Zoomen um sicherzustellen, dass es wirklich nicht funktioniert. Für die Aufnahme habe ich es aber etwas kürzer gestaltet.

Hier kann ich das Verhalten im Moment nicht nachstellen (FF & Chrome):
Bei zoom<=11 (3km) werden die farbigen Symbole angezeigt.
Für zoom>=12 (2km) werden die Cache-Icons korrekt angezeigt, auch beim reinzoomen.
Funktioniert das bei dir nicht so?

Das gif startet bei 5km und zoomt dann auf 3km. In beiden Fällen sieht man die Symbole, ich würde sagen das passt so.

Dass beim reinzoomen die Koordinaten im Suchfeld der Sidebar nicht aktualisiert werden ist korrekt, da ja keine neue Suche ausgeführt wird. Sondern nur dann, wenn notwendig.

2Abendsegler commented:
> Das gif startet bei 5km und zoomt dann auf 3km. In beiden Fällen sieht man die Symbole, ich würde sagen das passt so.
> 
> Dass beim reinzoomen die Koordinaten im Suchfeld der Sidebar nicht aktualisiert werden ist korrekt, da ja keine neue Suche ausgeführt wird. Sondern nur dann, wenn notwendig.

Hmm, es ist aber doch in diesem Fall notwendig. Würde ich die Suche nach dem hinein zoomen manuell anwählen, würden viele andere Caches angezeigt.
 
Zu Beginn sind die Symbole zentriert, es werden die maximale Anzahl von Symbolen in einem Kreis angezeigt. Ich bewege die Maus etwas nach rechts oben auf dem Bildschirm wo im Moment keine Symbole angezeigt werden und zoome hinein. Der Kreis mit den Symbolen wandert etwas nach rechts unten weg und der Kartenmittelpunkt ist nun auch ein anderer. Es ist quasi das Gleiche wie wenn ich die Karte etwas nach rechts unten schiebe, nur zoome ich gleichzeitig auch noch hinein. Ich habe damit eine Kartenbewegung durchgeführt und ich würde erwarten, dass nun eine neue Suche durchgeführt wird. Wenn das der Fall wäre, würden nun auch zentriert viele andere Symbole erscheinen. 

Tatsächlich funktioniert es ja auch manchmal, womöglich hängt das wieder mit den Errors zusammen. 

Wenn ich das Pluszeichen wähle, funktioniert das auch, obwohl ich damit nur hinein zoome und der Mittelpunkt der Karte sich nicht ändert. Oben und unten bewegt sich aber der Kreis mit den Symbolen genauso quasi über den Bildschirmrand hinweg und es verschwinden einige Symbole. In den Ecken des Bildschirms werden dann aber zusätzlich Symbole sichtbar, es wurde also eine neue Suche durchgeführt.

Ich würde vorschlagen, dass wir zuerst die Errors beheben. Danach passt das ja vielleicht. Ich werde mir das dann nochmal ansehen.

DieBatzen commented:
Die letzten 5% waren sehr hartnäckig, jetzt ist aber aus der 95%-Lösung eine 100%-Lösung geworden.

Ich habe versucht, nur das Nötigste zu ändern, aber selbst das war doch recht umfangreich.
Jetzt sieht alles gut aus, in all meinen Tests läuft das Feature nun rund und stabil - sowohl im FF als auch im Chrome.

Wenn du Zeit und Muße hast, kannst du dir den aktualisierten PR ja mal ansehen.

DieBatzen commented:
Während des Testens ist mir noch eine Ungenauigkeit nach dem Setzen der Default Filter aufgefallen:
die Anzahl der gefundenen Caches entsprach nicht denen, die im angezeigten Kartenausschnitt angezeigt werden.

Grund hierfür ist, dass das Setzen der Default Filter eine Radiussuche mit 16km Radius ausführt, auch wenn das Zoomlevel dazu nicht passt.
Mit einem `SearchThisArea` Aufruf nach dem Setzen der Default Filter stimmt dann die Anzahl der gefundenen Caches.

Den Fix habe ich in dem PR gleich mit eingebaut, da sich das einfach angeboten hat.

DieBatzen commented:
Das von dir beschriebene Verhalten beim Hineinzoomen sehe ich ebenfalls:
Maus nach rechts oben schieben, mit dem Mausrad hineinzoomen, es wird keine Suche durchgeführt.

Der Grund liegt in dieser Zeile:
`if (latHighG == false || latHigh > latHighG || latLow < latLowG || lngHigh > lngHighG || lngLow < lngLowG)`
Es wird geprüft, ob der neue Kartenausschnitt im alten enthalten ist. Ist dies der Fall, dann wird keine neue Suche ausgeführt.
Genau das ist bei dem oben beschriebenen Zoomvorgang der Fall.

Würde man erst die Karte verschieben und danach zoomen, dann sieht das anders aus und eine Suche wird ausgeführt (durch das Verschieben).

Da ließe sich aber bestimmt noch etwas optimieren ...

2Abendsegler commented:
Hört sich alles so weit gut an. Danke.
Ich versuche auch immer nur das Nötigste zu ändern, häufig ist das dann auch mehr als es anfangs zu erwarten war. Passt schon.
Bin noch bis Montag Abend unterwegs. Werde versuchen es am Montag Abend oder Dienstag anzugehen und nicht wieder ewig vor mir her zu schieben.

DieBatzen commented:
> > Das gif startet bei 5km und zoomt dann auf 3km. In beiden Fällen sieht man die Symbole, ich würde sagen das passt so.
> > Dass beim reinzoomen die Koordinaten im Suchfeld der Sidebar nicht aktualisiert werden ist korrekt, da ja keine neue Suche ausgeführt wird. Sondern nur dann, wenn notwendig.
> 
> Hmm, es ist aber doch in diesem Fall notwendig. Würde ich die Suche nach dem hinein zoomen manuell anwählen, würden viele andere Caches angezeigt.
> 
> Zu Beginn sind die Symbole zentriert, es werden die maximale Anzahl von Symbolen in einem Kreis angezeigt. Ich bewege die Maus etwas nach rechts oben auf dem Bildschirm wo im Moment keine Symbole angezeigt werden und zoome hinein. Der Kreis mit den Symbolen wandert etwas nach rechts unten weg und der Kartenmittelpunkt ist nun auch ein anderer. Es ist quasi das Gleiche wie wenn ich die Karte etwas nach rechts unten schiebe, nur zoome ich gleichzeitig auch noch hinein. Ich habe damit eine Kartenbewegung durchgeführt und ich würde erwarten, dass nun eine neue Suche durchgeführt wird. Wenn das der Fall wäre, würden nun auch zentriert viele andere Symbole erscheinen.
> 
> Tatsächlich funktioniert es ja auch manchmal, womöglich hängt das wieder mit den Errors zusammen.
> 
> Wenn ich das Pluszeichen wähle, funktioniert das auch, obwohl ich damit nur hinein zoome und der Mittelpunkt der Karte sich nicht ändert. Oben und unten bewegt sich aber der Kreis mit den Symbolen genauso quasi über den Bildschirmrand hinweg und es verschwinden einige Symbole. In den Ecken des Bildschirms werden dann aber zusätzlich Symbole sichtbar, es wurde also eine neue Suche durchgeführt.
> 
> Ich würde vorschlagen, dass wir zuerst die Errors beheben. Danach passt das ja vielleicht. Ich werde mir das dann nochmal ansehen.

Dafür habe ich nun eine Lösung gefunden, die ich, um Zeit zu sparen, vor Beginn der Tests noch schnell eingespielt habe:
- werden im aktuellen Kartenausschnitt bereits alle Suchergebnisse angezeigt  
![grafik](https://github.com/user-attachments/assets/a97c97b0-3858-4275-ab44-a4119631d6ea)  
dann muss beim hineinzoomen keine weitere Suche ausgeführt werden
  - das gilt solange, wie der Kartenausschnitt nicht verlassen wird
- anderenfalls  
![grafik](https://github.com/user-attachments/assets/d35783fa-ef75-4162-abed-8e3be16a70ff)  
werden beim hineinzoomen die Suchergebnisse aktualisiert

2Abendsegler commented:
Moin,

zuerst die schöne Sache. Auch bei mir funktionieren deine Änderungen von gestern Abend bezüglich der Aktualisierung der Suchergebnisse beim Hineinzoomen.   

Ich habe begonnen die neue Version hinsichtlich der Fehlermeldungen zu testen. Leider kommt es bei mir noch immer bei fast jeder Anwahl der Seite zum bekannten Error. Es kommt immer zeitversetzt. Man meint die Karte wäre bereits aufgebaut und dann kommt der Error. Das ist der neue Bereich im Coding mit den Verzögerungen. Es sind nach wie vor die beiden `click()` Zeilen im Bereich 
`// Observe distance value for changes, then run filtered search (only once).`
Also
`document.querySelector('button[data-event-label="Filters - Apply"]').click();`
und 
`document.querySelector('[data-testid="search-this-area-button"]').click();`
Beide sind etwa gleichermaßen für die Fehler bei mir verantwortlich. Ich habe die Seite 20 mal angewählt und erhalte 18 mal einen Fehler.

Ich habe setTimeout und beide setInterval auf 2 Sekunden erhöht, ohne dass ich geprüft habe, ob das überhaupt Sinn macht. Die Fehler entstehen immer noch. 

Womöglich trägt bei mir die eher langsame Internetverbindung zu den Fehlern bei.

Ich glaube nicht, dass die Verzögerungen zum Erfolg führen. Zudem führt das zu einer spürbaren Verlangsamung der Seite.

Vielleicht ist es der falsche Weg zu versuchen den Fehler zu vermeiden. Vielleicht wäre es sinnvoller das Problem mal bei Mozilla anzusprechen. Wie schätzt du das denn ein, ist das ein Fehler in Firefox? In Chrome funktioniert es ja. Und wärst du in der Lage das Problem zu beschreiben, so dass die das nachvollziehen können?

Ansonsten fällt mir im Moment nichts ein. Aus meiner Sicht sollten wir die Features `Defaults setzen` und `Zoom berücksichtigen` zurückbauen. Das sind die beiden Features die die Probleme verursachen, richtig?

DieBatzen commented:
Seltsam, dass das bei dir im FF so fehleranfällig ist. Bei mir läuft das auch im FF völlig problemlos.

Wenn du den zweiten Click `document.querySelector('[data-testid="search-this-area-button"]').click();` im `setInterval` auskommentierst, sind dann die Fehlermeldungen weg?

2Abendsegler commented:
> Wenn du den zweiten Click `document.querySelector('[data-testid="search-this-area-button"]').click();` im `setInterval` auskommentierst, sind dann die Fehlermeldungen weg?

Ja, dann scheinen die Fehlermeldungen nicht mehr aufzutauchen.
Wenn ich zusätzlich das `setTimeout(()` entferne dann kommt es in vielleicht 25% der Versuche wieder zum Fehler wegen der ersten Click Zeile.

DieBatzen commented:
Ok, dann liegt es daran, dass der zweite Click zu früh ausgeführt wird: falls die Filtersuche noch nicht vollständig beendet ist und die searchThisArea-Suche bereits anläuft, dann wird der Fehler ausgelöst.

Um genau das zu verhindern sind eigentlich die beiden `setInterval`-Aufrufe da.
Die Schwierigkeit ist hier, genau zu bestimmen, wann die erste Suche beendet ist.
Der Trigger sind die Zoom-Buttons, die werden während einer Suche deaktiviert.
Sobald der Zoom-Button wieder aktiviert wird, kann die zweite Suche starten.

Ich vermute, dass das in deinem Fall nicht wie erwartet funktioniert - wobei ich im Moment noch keine Idee habe, warum.

2Abendsegler commented:
Ok. Zumindest kann ich aber bei mir erkennen, dass die Zoom Buttons deaktiviert und wieder aktiviert werden.

DieBatzen commented:
> Wenn ich zusätzlich das setTimeout(() entferne dann kommt es in vielleicht 25% der Versuche wieder zum Fehler wegen der ersten Click Zeile.

Das `setTimeout()` ist notwendig um sicherzustellen, dass die Filter gesetzt sind, bevor die Filtersuche startet.
Von daher kommt der Fehler, oder auch nicht, wenn die Suche nicht zeitverzögert startet.

> Zumindest kann ich aber bei mir erkennen, dass die Zoom Buttons deaktiviert und wieder aktiviert werden.

Wenn es so läuft, wie es soll, dann solltest du zwei Mal (kurz hintereinander) sehen, dass die Zoom-Buttons deaktiviert und wieder aktiviert werden - für jede Suche einmal.

2Abendsegler commented:
> Wenn du den `document.querySelector('[data-testid="search-this-area-button"]').click();` zeitverögert ausführst, z.B. mit `setTimeout(() => {document.querySelector('[data-testid="search-this-area-button"]').click()}, 4000);`, ist der Fehler dann weg?

Bei 4000 funktioniert es. Bei 500 aber in einigen Fällen wieder nicht.

> Wenn es so läuft, wie es soll, dann solltest du zwei Mal (kurz hintereinander) sehen, dass die Zoom-Buttons deaktiviert und wieder aktiviert werden - für jede Suche einmal.

Es scheint so, dass es in einigen Fällen dazu kommt, dass die Zoom-Buttons drei mal deaktiviert werden. Als ich mit den 500 getestet habe waren das genau die Fälle, in denen es die Fehlermeldung gab. Reduziere ich die 500 weiter, kann man gut erkennen, dass das zweite Mal deaktivieren immer kürzer wird. Das ist wohl die Stelle wo es zum Fehler kommt weil sie abgebrochen wird durch den dritten Click. 

Übrigens kann ich diesen Fehler auch ohne GClh provozieren, indem ich fortwährend auf den Search Button klicke, der ja identisch deaktiviert wird zusammen mit den Zoom Buttons. 

Ich muss morgen früh raus und muss jetzt in die Heia. Bin dann vermutlich erst morgen Abend wieder erreichbar.

2Abendsegler commented:
Ich habe die letzten Tage noch einige Tests gemacht. Den Fehler gibt es tatsächlich nur dann, wenn die Zoom Buttons drei mal und nicht zwei mal deaktiviert und wieder aktiviert werden.  

Wenn ich den `document.querySelector('[data-testid="search-this-area-button"]').click();` mit 500 ms zeitverzögert ausführe, passierte es heute morgen etwa bei 10% der Seitenaufrufe. Die mittlere (zweite) Deaktivierung blinkt dabei immer nur kurz auf. Reduziere ich die ms weiter, passiert es immer häufiger, bis es ohne diese Verzögerung heute morgen bei etwa 25% der Seitenaufrufe passiert. Das Aufblinken wird dann auch umso kürzer.

Für mich sieht es so aus, als würde GS nach unserem Filter Apply und dessen Ausführung aus irgendwelchen Gründen die Suche noch mal starten. Lasse ich unseren Search Button Click weg, kommt es genauso oft vor, dass es zu dieser zweiten Deaktivierung kommt.  

Ich wollte eigentlich den Kommentar hier schon wegschicken, dann ist mir bei den letzten Tests aufgefallen, dass es jetzt bei jedem Seitenaufruf passiert. Ich habe den Browser geschlossen, hilft aber alles nichts. Passiert gerade bei jedem Seitenaufruf, diese zusätzliche Deaktivierung. Ich habe nichts an der Ausgangssituation geändert.

Noch eine Frage:
Du schreibst im Coding beim Filter Apply Click
```
  document.querySelector('button[data-event-label="Filters - Apply"]').click();
  // Run searchThisArea as soon as filtered search is finished
  // (filtered search sets the filters correct, but not the results according to zoom level).
```
Bedeutet das, dass wir den Search anschließend nur nochmal anstarten müssen weil wir den Zoom in der url verändert haben? Oder verstehe ich das falsch?

DieBatzen commented:
> Ich habe die letzten Tage noch einige Tests gemacht. Den Fehler gibt es tatsächlich nur dann, wenn die Zoom Buttons drei mal und nicht zwei mal deaktiviert und wieder aktiviert werden.
> 
> Wenn ich den `document.querySelector('[data-testid="search-this-area-button"]').click();` mit 500 ms zeitverzögert ausführe, passierte es heute morgen etwa bei 10% der Seitenaufrufe. Die mittlere (zweite) Deaktivierung blinkt dabei immer nur kurz auf. Reduziere ich die ms weiter, passiert es immer häufiger, bis es ohne diese Verzögerung heute morgen bei etwa 25% der Seitenaufrufe passiert. Das Aufblinken wird dann auch umso kürzer.
> 
> Für mich sieht es so aus, als würde GS nach unserem Filter Apply und dessen Ausführung aus irgendwelchen Gründen die Suche noch mal starten. Lasse ich unseren Search Button Click weg, kommt es genauso oft vor, dass es zu dieser zweiten Deaktivierung kommt.
> 
> Ich wollte eigentlich den Kommentar hier schon wegschicken, dann ist mir bei den letzten Tests aufgefallen, dass es jetzt bei jedem Seitenaufruf passiert. Ich habe den Browser geschlossen, hilft aber alles nichts. Passiert gerade bei jedem Seitenaufruf, diese zusätzliche Deaktivierung. Ich habe nichts an der Ausgangssituation geändert.
> 

Ich konnte das Verhalten unter FF nun endlich nachvollziehen und denke ich weiß, woran es liegt.
Die Filter-Suche `document.querySelector('button[data-event-label="Filters - Apply"]').click();` und SearchThisArea-Suche `document.querySelector('[data-testid="search-this-area-button"]').click();` kommen sich unter gewissen Umständen in die Quere.

Ich versuche das mal, etwas strukturiert zu erklären:
1. Default Filter aktiv, Corrected Coordinates inaktiv, automatische Suche nach Verschieben/Zoomen inaktiv
    - Um die Default Filter zu setzen, muss einmalig eine Filter-Suche durchgeführt werden
    - Diese Filter-Suche ist eine Radius-Suche und wird mit Radius=16km durchgeführt
    - Die Anzahl der gefundenen Caches passt aber in den meisten Fällen nicht zu dem angezeigten Kartenausschnitt, der durch das Zoom-Level in der URL (Parameter `zoom`) festgelegt wird, z.B. 418 Treffer bei der Radius-Suche mit Radius=16km vs. 90 Treffer bei zoom=14.
    - Um das zu kompensieren, wird direkt nach der Radius-Suche noch eine SearchThisArea-Suche durchgeführt, die nur Treffer aus dem angezeigten Kartenausschnitt liefert (Parameter `box` in der URL). Mehr dazu weiter unten.
2. Default Filter aktiv, Corrected Coordinates aktiv, automatische Suche nach Verschieben/Zoomen inaktiv
    - Ist Corrected Coordinates aktiv, dann muss einmalig nach dem Seitenaufbau eine Suche (Filter oder SearchThisArea) durchgeführt werden, um die Caches initial an der richtigen Position darzustellen.
    - Sind die Default Filter aktiv, wird bereits eine Suche getriggert und somit muss keine weitere Suche gestartet werden. Das wurde mit #2747 gefixt.
3. Default Filter aktiv, Corrected Coordinates aktiv, automatische Suche nach Verschieben/Zoomen aktiv
    - Ist die automatische Suche nach Verschieben/Zoomen aktiv, dann mus nach jedem Verschieben/Zoomen eine SearchThisArea-Suche durchgeführt werden.
    - Aber natürlich nur dann, wenn notwendig. Und da wird es komplex ...
        - Wir triggern das Verschieben/Zoomen durch die Änderung der URL
        - Leider ändert eine SearchThisArea-Suche selbst die URL, also muss diese potentielle zweite Suche verhindert werden
        - Genauso ändert eine Filter-Suche die URL, auch hier muss eine potentielle zweite Suche verhindert werden
    - Dann gibt es noch die Sonderfälle, wo keine Default Filter gesetzt werden (z.B. Matrix-Suche) und daher keine Filter-Suche durchgeführt wird. Trotzdem muss dann für die Corrected Coordinates noch eine Suche durchgeführt werden
    - usw.

Auch wenn scheinbar alles gut aufeinander abgestimmt ist, gibt es immer noch Situationen, in denen zwei Suchen zeitlich zu nahe beieinander liegen. Und genau dann wird der Fehler ausgelöst.
Ich habe da aber schon eine Idee, wie man sicherstellen kann, dass das nicht mehr passiert.
Werde ich mir anschauen.

> Noch eine Frage: Du schreibst im Coding beim Filter Apply Click
> 
> ```
>   document.querySelector('button[data-event-label="Filters - Apply"]').click();
>   // Run searchThisArea as soon as filtered search is finished
>   // (filtered search sets the filters correct, but not the results according to zoom level).
> ```
> 
> Bedeutet das, dass wir den Search anschließend nur nochmal anstarten müssen weil wir den Zoom in der url verändert haben? Oder verstehe ich das falsch?

Das habe ich versucht bei 1. zu erläutern (hoffentlich verständlich).
Beim Durchsehen des Codes kam mir dann noch eine Idee, die diese zweite Suche überflüssig macht.
Wir können der Filter-Suche frühzeitig einen dem Zoom-Level passenden Suchradius mitgeben, dann passt die Anzahl der Treffer direkt.

Wenn ich eine neue Version fertig habe, dann aktualisiere den PR - dann hoffentlich in einer finalen Version.

DieBatzen commented:
Alles sehr gute Punkte, ich fange mal hinten an (mit der schlechten Nachricht) ...
> 
> Können wir nicht vielleicht versuchen die Spezifikationen in die url einbauen, so wie es beim zoom gemacht wird, wenn der zoom von GS überschrieben wird? Found status, Owner status und Geocache types werden ja auch in der url abgebildet nachdem sie im Filter verwendet wurden. Könnten wir damit nicht erreichen, dass GS die Suche selbst organisiert und wir die Buttons gar nicht betätigen müssen und auch keine Wartezeiten einbauen müssen?
>
Das ist genau die richtige Idee und hatte ich zuallererst versucht.
Aber leider funktioniert das technisch nicht.

Denn wird die Seite geladen, dann sind bereits alle Suchergebnisse, inkl. Suchparameter, in der Seite (als Script) eingebettet und werden so später verarbeitet.
Das Ändern der URL bzw. deren Parameter hat darauf leider keinen Einfluss, da standardmäßig keine weitere Abfrage nach Hause erfolgt.

Die einzige Möglichkeit, die Suchergebnisse zu aktualisieren, ist so früh wie möglich eine Filtersuche anzustoßen (unser Ansatz).
Für diese Filtersuche ändern wir vorher noch die Default Filter und, falls notwendig, den Suchradius. Dann sind die Suchergebnisse wie erwartet.

Beim Ändern des Zoom-Levels gehen wir anders vor, indem die Leaflet-Funktion `setZoom()` aufgerufen wird.
Das ändert dann auch den zoom-Parameter der URL.

> 
> Aus meiner Sicht gibt es zwei grundlegende Probleme.
> 
> Zum einen benötigt die Seite einfach zu lange, bis der User mit ihr arbeiten kann. Und es ist nervig, wenn man die Karte schon verschieben oder zoomen kann und dann durch unsere Aktionen wieder auf den Ursprungsstand zurück befördert wird.
>
Das ist natürlich völlig richtig, die notwendigen Verzögerungen sollten minimal sein.
Aber mehr dazu weiter unten (die gute Nachricht).

>
> Und zum anderen können wir nicht sicherstellen, dass es für alle User passt und es nicht zum Erlahmen des Firefox Browsers kommt, weil beim User noch einmal eine leicht andere Konstellation vorhanden ist. Auch deshalb weil GS hier jederzeit wieder rumbasteln kann und dadurch unsere Annahmen beispielsweise mit der Wartezeit nach dem Klick wieder völlig über den Haufen geworfen werden können. Und hier handelt es sich nicht um einen Fehler, der dazu führt, dass ein Feature oder auch mehrere Features vom GClh nicht mehr funktionieren. Wir könnten vielmehr dafür sorgen, dass der Browser für den ein oder anderen User nicht mehr nutzbar ist, so wie bei mir quasi der Fall, wenn ich die Seite im Verlauf mehrfach anwähle. Es wird nicht jeder User feststellen können, woran es liegt. Indem wir versuchen an der derzeitigen Vorgehensweise festzuhalten, provozieren wir aus meiner Sicht solche Situationen. Das finde ich nicht gut.
> 
Auch da hast du völlig recht, der Browser darf natürlich nicht beeinträchtigt werden.
Und Timeouts einzubauen ist immer unschön. Allein, wenn die Webseite mal viel langsamer als üblich reagiert, kann das noch funktionieren ... oder auch nicht.
Auch hier mehr dazu weiter unten (die gute Nachricht).

>
> Ich sehe im Moment keine Lösung für diese beiden Probleme, solange wir versuchen alles über Klicks auf Buttons abzubilden.
>
Nun aber zur guten Nachricht ...
Da bin ich optimistischer, denn ich denke mittlerweile beide Probleme gelöst zu haben, insbesondere für den FF. Alle meine Tests zeigen keine Auffälligkeiten mehr.

Im Code wird nun ausgeschlossen, dass mehr als eine Suche gleichzeitig starten bzw. laufen kann. Dadurch kann der Fehler, der die Probleme verursacht hat, nicht mehr auftreten und der FF lahmt auch nicht mehr.

Die leichte Verzögerung (jetzt kürzer als vorher), die wir beim initialen Laden der Seite verursachen (müssen), finde ich verschmerzbar - aber das ist natürlich eine subjektive Einschätzung.

Ich aktualisiere den PR mit den erwähnten Änderungen und setze den PR vorerst auf Draft.

Für das Setzen der Default-Filter können wir wahrscheinlich auf die bisherigen Klicks verzichten und das analog zum Setzen des Suchradius durchführen. Das werde ich mir noch anschauen, sieht aber schon ziemlich gut aus.

DieBatzen commented:
> Für das Setzen der Default-Filter können wir wahrscheinlich auf die bisherigen Klicks verzichten und das analog zum Setzen des Suchradius durchführen. Das werde ich mir noch anschauen, sieht aber schon ziemlich gut aus.
>
Das ist nun im letzten Commit enthalten: die Default Filter werden nicht mehr per Klick im Filterdialog, sondern zum frühst möglichen Zeitpunkt durch Anpassen der Filterparameter gesetzt.
Das macht es insgesamt etwas robuster und auch minimal schneller.

Weitere Möglichkeiten für Verbesserungen oder alternative Ansätze sehe ich nicht mehr.

Falls der Code bei dir unter FF immer noch Schwierigkeiten macht, dann bleibt wohl nur noch eine Option:
alle Features deaktivieren, die Probleme machen.

Aber ich bin optimistisch, dass es nun auch im FF stabil läuft.

DieBatzen commented:
Ich habe jetzt doch noch eine letzte Idee verfolgt und tatsächlich eine andere Lösung gefunden, die alle bisherigen Probleme bereits im Ansatz behebt.
Damit funktionieren sowohl das Setzen der Default Filter, als auch das automatische Aktualisieren der Suchergebnisse bei Verschieben/Zoomen der Karte problemlos (FF und Chrome).
Zusätzlich stimmt nun die Anzahl der gefundenen Caches nach dem Laden der Seite und die Caches werden direkt an den korrigierten Koordinaten angezeigt (falls aktiviert).
Dafür sind insbesondere keine Filtersuchen mehr notwendig.

Die Grundidee ist, alle notwendigen Anpassungen durchzuführen, bevor die eigentliche Verarbeitung der Suchergebnisse beginnt.
Das war bisher nicht möglich, da beim Start des GCLH diese Verarbeitung bereits beendet war (Stichwort `@run-at document-end`).

Über den Tampermonkey-Parameter `@run-at document-start` lässt sich der GCLH aber so früh starten, dass es doch möglich ist.
In Pseudocode sieht das so aus:
`@run-at document-start`
`- Caches an korrigierte Koordinaten setzen (falls aktiviert)` 
`- Default Filter über die URL setzen`
`- Seite mit modifizierter URL neu laden`
`@run-at document-end`
`- Den GCLH einfach wie immer ausführen, keine Änderungen notwendig`

Da außer dem neuen Code der GCLH immer noch bei `@run-at document-end` startet, sind alle Funktionen durch den prinzipiell früheren Start des Skripts nicht betroffen.

Zukünftig könnte man dieses Setup sogar bei ähnlich gelagerten Problemen genauso verwenden.

Ich aktualisiere den PR und setze ihn wieder frei zum Review.

2Abendsegler commented:
@DieBatzen 
Moin, 

nur weil ich gerade hier etwas zu tun hatte, weil capoaira ein neues Problem im #2708 mit der Search Map gemeldet hat, schreibe ich mal kurz was zu den letzten Änderungen. 

Bei mir laufen die Änderungen wieder gegen die Wand. Error in Zeile 193 `observer.observe(target, config);`. Anschließend wird die Seite immer wieder neu aufgerufen. Womöglich hängt letzteres mit dem Error zusammen. 

Unabhängig von diesen Problemen mit den Änderungen, zweifle ich etwas daran, dass wir solche drastischen Änderungen implementieren sollten. Tatsächlich bin ich schon einige Tage am Überlegen wie wir hiermit weiter verfahren sollen und komme irgendwie zu keinem zufriedenstellenden Ergebnis. Dabei geht es auch nicht nur um die aktuellen Anpassungen, sondern auch darum, ob das Ganze in Zukunft noch wartbar ist. 

Ich brauche noch etwas Zeit, um mir hoffentlich etwas mehr Klarheit zu verschaffen. 🙁

DieBatzen commented:
> Bei mir laufen die Änderungen wieder gegen die Wand. Error in Zeile 193 `observer.observe(target, config);`. Anschließend wird die Seite immer wieder neu aufgerufen. Womöglich hängt letzteres mit dem Error zusammen.

Das ist seltsam und dürfte, wie bei dem Linux-Fehler, eigentlich gar nicht passieren. Aber es passiert ja trotzdem ...

Die einzige Möglichkeit wäre, dass das `__NEXT_DATA__`-Skript zu diesem Zeitpunkt noch nicht vorhanden ist.
Gut, auch das lässt sich einfach fixen, indem sichergestellt wird, dass es vor der Code-Ausführung vorhanden ist.

Ich habe den PR noch einmal aktualisiert, nun müsste es aber klappen.

DieBatzen commented:
> Unabhängig von diesen Problemen mit den Änderungen, zweifle ich etwas daran, dass wir solche drastischen Änderungen implementieren sollten. Tatsächlich bin ich schon einige Tage am Überlegen wie wir hiermit weiter verfahren sollen und komme irgendwie zu keinem zufriedenstellenden Ergebnis. Dabei geht es auch nicht nur um die aktuellen Anpassungen, sondern auch darum, ob das Ganze in Zukunft noch wartbar ist.
> 
> Ich brauche noch etwas Zeit, um mir hoffentlich etwas mehr Klarheit zu verschaffen. 🙁

Mit der Wartbarkeit des Codes hast du natürlich einen guten Punkt.
Wir machen das ja alle in unserer Freizeit und weil wir es gerne tun, da möchte man nicht Stunde um Stunde mit testen verbringen.

Ich denke, der wesentliche Punkt für aktuelle und zukünftige Entwicklungen ist sicherzustellen, dass Abhängigkeiten (Variablen, DOM-Elemente) vorhanden sind, bevor ein darauf basierender Code ausgeführt wird. Speziell mit dem Tech Update, wo vieles verzögert geladen wird.
Das haben die letzten Probleme mit FF vs Chrome oder Windows vs Linux gezeigt - Browser und Betriebssysteme verhalten sich leider nicht immer gleich - mal wird schneller geladen, mal langsamer.

Die Änderungen in diesem PR sehe ich eigentlich gar nicht so drastisch.
Bis auf den neuen Code, der so früh wie möglich am Anfang des Skripts ausgeführt wird, startet der Rest des Skripts wie immer - gekapselt in dem `@run-at document-end` entsprechenden Eventlistener.

Soweit ein paar Gedanken meinerseits.

2Abendsegler commented:
@DieBatzen 
Danke für deine Gedanken!

Wie ich schon mehrfach angedeutet habe, bin ich nicht mehr optimistisch, dass wir die Startsequenz in der Search Map in den Griff bekommen.

Im Vordergrund steht bei mir immer, dass wir einen Mehrwert generieren. Der Aufwand für die Generierung dieses Mehrwertes muss dabei im Verhältnis bleiben. Dazu ist es auch notwendig irgendwann mit den Arbeiten ein Ende zu finden. Idealerweise mit der Herausgabe des Mehrwertes an die Nutzer. Wenn man bei der Herausgabe an die Nutzer schon davon ausgehen muss, dass das Thema immer wieder bei einem aufschlägt, dann ist die Herausgabe nicht der richtige Weg. 

Bei der Startsequenz in der Search Map und beim Umbau im #2739, hangeln wir uns gefühlt nur von einem Fehler zum anderen. Wobei es nur zufällig bei mir zu den Fehlern kommt, weil ich zufällig eben diese, meine Konstellation habe. Bei anderen Nutzern kommt es womöglich nochmal zu ganz anderen Fehlern. Grundsätzlich ist das ja immer so. Im Idealfall hat man aber im Vorfeld bereits alle Eventualitäten berücksichtigt.

Hier scheint das aber nicht möglich alle Eventualitäten zu berücksichtigen. Das Umfeld scheint einfach zu komplex und zu anfällig für Probleme. Es scheint nicht so, dass wir dieses Umfeld im Griff haben. Für mich sieht es so aus, als ob wir ziemlich viel herumstochern, wir uns also immer noch in einem Lernprozess befinden was die Umgebung angeht. Und leider sieht es nicht so aus, als wäre dabei ein Ende in Sicht, auch wenn wir schon viele Schritte weitergekommen sind. Bitte verstehe mich nicht falsch, das ist kein Vorwurf an dich. Du hast super Arbeit geleistet und hast dich klasse eingebracht. Das Problem liegt in der komplexen und unübersichtlichen Umgebung.

Ich finde deine letzten Änderungen sehr wohl drastisch. Eine geniale Idee, keine Frage, aber drastisch sind sie. 

Indem du die anfängliche Verarbeitung der Search Map ganz vorne anstellst, werden einige über die Jahre entstandenen, etablierten Verfahren, die zu Beginn der Verarbeitung im Script dafür sorgen, dass alles in richtige Bahnen kommt, einfach ausgehebelt. 

Beispielsweise sind die Config Parameter noch nicht vorhanden, deshalb musst du sie selbst besorgen. Das Arbeiten mit try-catch im etablierten Verfahren mit der Fehlerbehandlung über "function gclh_error" ist noch nicht möglich. Überhaupt wurde noch gar nicht geprüft, ob wir uns auf einer der Geocaching Seiten befinden. Enthält die url ein "/play/map" würde der neue Programmteil auch bei dieser url starten. Würde das Script unter Greasemonkey gestartet, würde der neue Programmteil auch ausgeführt, weil ein Stop wegen Greasemonkey erst später kommt. Unsere jQuery Version ist wohl auch noch nicht verfügbar, sondern wenn überhaupt nur die Version der aktuellen Webseite. Die Ausgabe von Informationen zum Stand der Verarbeitung über "function tlc" wird hier auch nicht unterstützt, kann aber ein wichtiger Baustein sein um zu prüfen an welcher Stelle es beim User hakt und was noch ok gelaufen ist. Vermutlich läuft der neue Programmteil auch dann wenn gar kein User angeloggt ist.

Das sind nur mal die Dinge, die mir auf Anhieb einfallen, das ist aber sicherlich nicht vollständig. 

Es geht jetzt nicht darum für alle diese Punkte eine Lösung zu finden und nachzubessern. Der Aufwand übersteigt schon jetzt den erdachten Mehrwert deutlich. Und die Arbeiten machen mir keine Freude mehr.

Zudem können wir aus meiner Sicht auch mit deinen letzten Änderungen nicht sicherstellen, dass wir bei allen Konstellationen immer in der Lage sind rechtzeitig die url anzupassen und die Seite neu zu laden.

Niemand von uns kann wirklich abschätzen, wie die Auswirkungen im Detail bei irgendwelchen Konstellationen aussehen könnten. Ich möchte eigentlich ein solches Risiko lieber nicht eingehen. Und dass es ein Risiko ist, haben wir in den vergangen Wochen zahlreiche Male gesehen.   
  
Grundsätzlich möchte ich keine Änderungen an die Nutzer geben, wenn ich nicht nahezu ausschließen kann, dass es bei auch nur wenigen Nutzern gegen die Wand läuft. Wären bei mir die Probleme nicht zufällig aufgetaucht, weil ich eben diese, meine Konstellation habe, hätten wir schon mehrfach Änderungen rausgegeben, die genau dazu geführt hätten. Keine Frage, wir können auch Fehler machen, aber nicht in Folge immer wieder. 

Eigentlich sollten wir einige der problematischeren Programmteile zurückbauen für die wir einen neuen Search durchführen müssten. Und wir sollten versuchen ein paar Vereinfachungen zu machen. Womöglich sollten wir auch zusätzliche Config Parameter vorsehen, was wir ziemlich vernachlässigt haben, um irgendwelche Features im Rahmen der Startsequenz abzuschalten, wenn es dort zu Problemen kommt. 

Ich weiß jetzt nicht wie wir das organisieren sollen, ist auch schon wieder spät geworden. Im #2739 könnten wir "function searchThisArea" behandeln, das wollen wir ja weiter einbauen. In einem neuen PR könnten wir - naja, genau genommen eher du - den Rückbau betreiben und gegebenenfalls Vereinfachungen und neue Config Parameter einbauen. Beim #2786 sollte vielleicht auch noch nachgebessert werden. Oder wir benennen den #2739 um, weil wir hier das Meiste zu den Problemen stehen haben und betreiben den Rückbau hier. Für "function searchThisArea" könnten wir dann einen anderen PR aufmachen.

DieBatzen commented:
> Eigentlich sollten wir einige der problematischeren Programmteile zurückbauen für die wir einen neuen Search durchführen müssten.

Ja, das scheint der richtige Weg zu sein.
Insbesondere das Setzen der Default Filter ist letztendlich die Ursache fast aller Probleme.
Dieser Code ist recht komplex und hat diverse Abhängigkeiten und Spezialfälle, die auch andere Funktionen beeinflussen können.
Den Mehrwert dieser Funktion in der Search Map sehe ich hingegen als überschaubar. Man kann ja auch jederzeit Such-Links im Browser hinterlegen, die jede gewünschte Suche abspeichern.
Für die Browse Map hat das natürlich seine Berechtigung.

Für das weitere Vorgehen würde ich folgendes vorschlagen:

1. Wir schließen diesen PR, damit bleibt zumindest der Code mit seinen Änderungen erhalten. Einiges kann man sicher wiederverwenden.
2. Die Default Filter in der Search Map werden nicht mehr unterstützt, der zugehörige Code entfernt, die Config entsprechend angepasst. Das passiert in einem eigenen PR.
3. Ist 2. erledigt, dann in einem weiteren PR `function searchThisArea` reaktivieren.

Im Anschluss kann dann auch #2708 geschlossen werden.
Insgesamt sollte das wieder zu einer Codebasis führen, die robust funktioniert und wartbar bleibt.

2Abendsegler commented:
> Für das weitere Vorgehen würde ich folgendes vorschlagen:
>     1. Wir schließen diesen PR, damit bleibt zumindest der Code mit seinen Änderungen erhalten. Einiges kann man sicher wiederverwenden.
>     2. Die Default Filter in der Search Map werden nicht mehr unterstützt, der zugehörige Code entfernt, die Config entsprechend angepasst. Das passiert in einem eigenen PR.
>     3. Ist 2. erledigt, dann in einem weiteren PR `function searchThisArea` reaktivieren.

Ja, prima, danke, lass es uns so machen.

> Im Anschluss kann dann auch #2708 geschlossen werden. Insgesamt sollte das wieder zu einer Codebasis führen, die robust funktioniert und wartbar bleibt.

Ich muss noch einiges zum Compact Layout einbauen, das ist auch noch vor dem Schließen des #2708 zu erledigen.

DieBatzen commented:
Wie besprochen machen wir hier zu.
